boa/src/syntax/lexer/number.rs

```diff
@@ -0,0 +1,334 @@
+        }
+    }
+
+    /// Converts `self` to BigInt kind.
+    fn to_bigint(self) -> Self {
+        match self {
+            Self::Rational => unreachable!("can not convert rational number to BigInt"),
+            Self::Integer(base) => Self::BigInt(base),
+            Self::BigInt(base) => Self::BigInt(base),
+        }
+    }
+}
+
+fn take_signed_integer<R>(
+    buf: &mut String,
+    cursor: &mut Cursor<R>,
+    kind: &NumericKind,
+) -> Result<(), Error>
+where
+    R: Read,
+{
+    // The next part must be SignedInteger.
+    // This is optionally a '+' or '-' followed by 1 or more DecimalDigits.
+    match cursor.next()? {
+        Some('+') => {
+            buf.push('+');
+            if !cursor.next_is_pred(&|c: char| c.is_digit(kind.base()))? {
+                // A digit must follow the + or - symbol.
+                return Err(Error::syntax("No digit found after + symbol"));
+            }
+        }
+        Some('-') => {
+            buf.push('-');
+            if !cursor.next_is_pred(&|c: char| c.is_digit(kind.base()))? {
+                // A digit must follow the + or - symbol.
+                return Err(Error::syntax("No digit found after - symbol"));
+            }
+        }
+        Some(c) if c.is_digit(kind.base()) => buf.push(c),
+        Some(c) => {
+            return Err(Error::syntax(format!(
+                "When lexing exponential value found unexpected char: '{}'",
+                c
+            )));
+        }
+        None => {
+            return Err(Error::syntax("Abrupt end: No exponential value found"));
+        }
+    }
+
+    // Consume the decimal digits.
+    cursor.take_until_pred(buf, &|c: char| c.is_digit(kind.base()))?;
+
+    Ok(())
+}
+
+/// Utility function for checking the NumericLiteral is not followed by an `IdentifierStart` or `DecimalDigit` character.
+///
+/// More information:
+///  - [ECMAScript Specification][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-literals-numeric-literals
+fn check_after_numeric_literal<R>(cursor: &mut Cursor<R>) -> Result<(), Error>
+where
+    R: Read,
+{
+    let pred = |ch: char| ch.is_ascii_alphabetic() || ch == '$' || ch == '_' || ch.is_ascii_digit();
+    if cursor.next_is_pred(&pred)? {
+        Err(Error::syntax("NumericLiteral token must not be followed by IdentifierStart nor DecimalDigit characters"))
+    } else {
+        Ok(())
+    }
+}
+
+impl<R> Tokenizer<R> for NumberLiteral {
+    fn lex(&mut self, cursor: &mut Cursor<R>, start_pos: Position) -> Result<Token, Error>
+    where
+        R: Read,
+    {
+        let mut buf = self.init.to_string();
+
+        // Default assume the number is a base 10 integer.
+        let mut kind = NumericKind::Integer(10);
+
+        let c = cursor.peek();
+
+        if self.init == '0' {
+            if let Some(ch) = c? {
+                match ch {
+                    'x' | 'X' => {
+                        // Remove the initial '0' from buffer.
+                        cursor.next();
```

Comment:
we should handle this, because the `next` method can fail we should propagate it, when there is an io error
```suggestion
                        cursor.next()?;
```